src/clj/rems/event_notification.clj

```diff
@@ -0,0 +1,74 @@
+(ns rems.event-notification
+  (:require [clj-http.client :as http]
+            [clj-time.core :as time]
+            [clojure.test :refer :all]
+            [clojure.tools.logging :as log]
+            [mount.core :as mount]
+            [rems.api.schema :as schema]
+            [rems.config]
+            [rems.db.applications :as applications]
+            [rems.db.outbox :as outbox]
+            [rems.json :as json]
+            [rems.scheduler :as scheduler]
+            [rems.util :refer [getx]]))
+
+(def ^:private default-timeout 60)
+
+(defn- notify! [target body]
+  (log/info "Sending event notification to" (:url target))
+  (try
+    (let [timeout-ms (* 1000 (get target :timeout default-timeout))
+          response (http/put (getx target :url)
+                             {:body body
+                              :throw-exceptions false
+                              :content-type :json
+                              :headers (get target :headers)
+                              :socket-timeout timeout-ms
+                              :conn-timeout timeout-ms})
+          status (:status response)]
+      (when-not (= 200 status)
+        (log/error "Event notification response status" status)
+        (str "failed: " status)))
+    (catch Exception e
+      (log/error "Event notification failed" e)
+      "failed: exception")))
+
+(defn process-outbox! []
+  (doseq [entry (outbox/get-entries {:type :event-notification :due-now? true})]
+    (if-let [error (notify! (get-in entry [:outbox/event-notification :target])
+                            (get-in entry [:outbox/event-notification :body]))]
+      (let [entry (outbox/attempt-failed! entry error)]
+        (when (not (:outbox/next-attempt entry))
+          (log/warn "all attempts to send event notification id " (:outbox/id entry) "failed")))
+      (outbox/attempt-succeeded! (:outbox/id entry)))))
+
+(mount/defstate event-notification-poller
+  :start (scheduler/start! process-outbox! (.toStandardDuration (time/seconds 10)))
+  :stop (scheduler/stop! event-notification-poller))
+
+(defn- add-to-outbox! [target body]
+  (outbox/put! {:outbox/type :event-notification
+                :outbox/deadline (time/plus (time/now) (time/days 1)) ;; hardcoded for now
+                :outbox/event-notification {:target target
+                                            :body body}}))
+
+(defn wants? [target event]
+  (let [whitelist (:event-types target)]
+    (or (empty? whitelist)
+        (some? (some #{(:event/type event)} whitelist)))))
```

Comment:
can't return the first match because of the empty whitelist case

Reply:
I was thinking more of refactoring the boolean part of `wants?` a little bit. it can definitely handle the empty case, that's ok, I'm talking more of the unnecessary boolean logic here when it is in fact used only below in a `when`. Just wondering if the `?` is a guaranteed boolean convention is all in all useful.

Reply:
I just think the `(wants? target event)` reads best. Also, "does this target accept this event" seems like a nice piece of business logic to factour out and test separately.